minetest/minetest
# Extend bone override capabilities

Revival of #10704 - rebased, loads of bugs fixed, code cleaned up. Closes #12294. Fixes the bone rotation interpolation aspect of #10572. Closes #13688

Implements bone overrides supporting *position, rotation and scale*, all of which can be *relative and/or interpolated*.

Ready for review. Extensive testing needed. 

# How to test

1. Use mods like character_anim or playeranim to verify that the old API still works.

2. Use the new, documented API. Example based on @LoneWolfHT's headanim:
    1. Use the `/headanim` chatcommand to toggle the override.

![Screenshot 2022-05-30 14-49-23](https://user-images.githubusercontent.com/34514239/170999166-154a55a1-2cf9-437f-9f43-163428b8093d.png)

**`mods/headlessanim/init.lua`:**
```lua
local lastdir = {}

minetest.register_globalstep(function(dtime)
	for _, player in pairs(minetest.get_connected_players()) do
		local pname = player:get_player_name()
		local ldeg = -math.deg(player:get_look_vertical())
		if lastdir[pname] == nil then
			lastdir[pname] = 0
		end
		if lastdir[pname] then
			if math.abs(lastdir[pname] - ldeg) > 4 then
				lastdir[pname] = ldeg
				player:set_bone_override("Head", {rotation = {vector = {x = ldeg, y = 0, z = 0}, absolute = true },
				scale = {vector = { x = 1.5, y = 1.5, z = 1.5 }}})
			end
		end
	end
end)

minetest.register_on_leaveplayer(function(player)
	lastdir[player:get_player_name()] = nil
end)

minetest.register_chatcommand("headanim", {
	func = function(name)
		local player = assert(minetest.get_player_by_name(name))
		if lastdir[name] then
			lastdir[name] = false -- don't update Head in globalstep
			player:set_bone_override"Head" -- clear override
		else
			lastdir[name] = 0
		end
	end
})
```
